Couple of improvements and fixes

azizk opened the pull request:
Hi!

I'm using Elixir every moment at work and would love to see improvements to the syntax highlighting definition. That's why I'm contributing these changes.

I hope to add a syntax test file soon in order to ensure a better quality and to avoid possible regressions when we modify the regular expressions.

Besides, what do you think about adopting the style of other syntax definition files like the one for Python in the PythonImproved package? Maybe that's the path we should take to make ours better and more ideal?

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -5,11 +5,11 @@ name: Elixir
 
 variables:
-  module_name: '\b[A-Z]\w*\b'
+  module_name: '\b[A-Z][a-zA-Z0-9_]*\b'
```
`\w` is exactly that

```elixir
iex(3)> "0" =~ ~r/\w/
true
```

azizk replied:
Not quite, because it also matches Unicode characters like "á". The compiler rejects those and only allows ASCII characters in module names. "\w" is a nice shorthand but it does more than ASCII letters.

princemaple replied:
👍

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -335,13 +335,13 @@ contexts:
         - match: '{{module_name}}'
           scope: entity.name.protocol.elixir
-    - match: '^\s*(def|defmacro)\s+([a-zA-Z_]\w*(?:!|\?)?)(?:(\()|\s*)'
+    - match: '^\s*(def|defmacro)\s+(\w+(?:!|\?)?)(?:(\()|\s*)'
```
identifiers cannot start with numbers. see my previous comment, `\w` can match a number

azizk replied:
Oh, that's true. So it's a mistake. This should do:

```suggestion
    - match: '^\s*(def|defmacro)\s+([_[:alpha:]]\w*(?:!|\?)?)(?:(\()|\s*)'
```

princemaple replied:
Just tested, functions can actually start with unicode (or be entirely a unicode name).

princemaple replied:
Does [:alpha:] cover unicode?

azizk replied:
Yep, it does :+1:

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -335,13 +335,13 @@ contexts:
         3: punctuation.definition.parameters.elixir
       push: function_body
-    - match: '^\s*(defp|defmacrop)\s+([a-zA-Z_]\w*(?:!|\?)?)(?:(\()|\s*)'
+    - match: '^\s*(defp|defmacrop)\s+(\w+(?:!|\?)?)(?:(\()|\s*)'
```
same as above

azizk replied:
```suggestion
    - match: '^\s*(defp|defmacrop)\s+([_[:alpha:]]\w*(?:!|\?)?)(?:(\()|\s*)'
```

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -59,7 +59,7 @@ contexts:
           pop: true
         - include: main
-    - match: \b(is_atom|is_binary|is_bitstring|is_boolean|is_float|is_function|is_integer|is_list|is_map|is_nil|is_number|is_pid|is_port|is_record|is_reference|is_tuple|is_exception|abs|bit_size|byte_size|div|elem|hd|length|map_size|node|rem|round|tl|trunc|tuple_size)\b
+    - match: \b(is_(?:atom|binary|bitstring|boolean|float|function|integer|list|map|nil|number|pid|port|record|reference|tuple|exception)|abs|bit_size|byte_size|div|elem|hd|length|map_size|node|rem|round|tl|trunc|tuple_size)\b
```
👍

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -440,7 +440,7 @@ contexts:
       comment: as above, just doesn't need a 'end' and does a logic operation
       scope: keyword.operator.elixir
-    - match: '{{module_name}}'
+    - match: '{{module_name}}(?!:)'
```
when would you do that?

azizk replied:
When you have a map or list containing keys starting with a capital letter: `[A: 0, B: 1]`

princemaple replied:
That seems rare, but OK :)

princemaple commented on Elixir.sublime-syntax:
```diff
@@ -480,12 +480,12 @@ contexts:
         - include: interpolated_elixir
         - include: escaped_char
-    - match: '(?<!:)(:)(?>[a-zA-Z_][\w@]*(?>[?!]|=(?![>=]))?|\<\>|===?|!==?|<<>>|<<<|>>>|~~~|::|<\-|\|>|=>|=~|=|/|\\\\|\*\*?|\.\.?\.?|>=?|<=?|&&?&?|\+\+?|\-\-?|\|\|?\|?|\!|@|\%?\{\}|%|\[\]|\^(\^\^)?)'
+    - match: '(?<!:)(:)(?>[_[:alpha:]][\w@]*(?>[?!]|=(?![>=]))?|\<\>|===?|!==?|<<>>|<<<|>>>|~~~|::|<\-|\|>|=>|=~|=|/|\\\\|\*\*?|\.\.?\.?|>=?|<=?|&&?&?|\+\+?|\-\-?|\|\|?\|?|\!|@|\%?\{\}|%|\[\]|\^(\^\^)?)'
```
what does this add？

azizk replied:
Allows Unicode atoms like `:á` or even `:_`

princemaple replied:
guess that answers my question above :)

princemaple commented:
> Besides, what do you think about adopting the style of other syntax definition files like the one for Python in the PythonImproved package? Maybe that's the path we should take to make ours better and more ideal?

What do you mean exactly? I'm not quite sure what PythonImproved does that you are referring to specifically. Some examples might help.

FYI I have no intention messing with tmLanguage in my lifetime 😆 .
There is https://github.com/elixir-editors/elixir-tmbundle for that.

princemaple commented:
Hey @azizk , looks like you did not tick the allow maintainer to modify your code checkbox. You will have to commit the suggestions yourself. They look good 👍 .

azizk commented:
Thanks for the review! I added my suggestions.

Of course, I don't want any tmLanguage XML craziness either. The PythonImproved package uses YAML. It looks very extensive, complete and well thought out. I don't know Sublime's syntax file format fully yet, but one main difference to Elixir.sublime-syntax is that PythonImproved mainly uses beginCaptures and endCaptures which may enable Sublime to actually parse the text into a proper syntax tree for more accurate and contextful highlighting, which is better than just having some semi-tree structure with flat areas where regexes match blindly more or less. It's fine in the beginning but it can definitely be improved.

azizk commented:
Oh, but I think I did tick the box to allow changes by you. Don't know why it's not working.

princemaple commented:
> I'll take a look again tomorrow. Until then!

Thank you very much!

> Of course, I don't want any tmLanguage XML craziness either. The PythonImproved package uses YAML. It looks very extensive, complete and well thought out. I don't know Sublime's syntax file format fully yet, but one main difference to Elixir.sublime-syntax is that PythonImproved mainly uses beginCaptures and endCaptures which may enable Sublime to actually parse the text into a proper syntax tree for more accurate and contextful highlighting, which is better than just having some semi-tree structure with flat areas where regexes match blindly more or less. It's fine in the beginning but it can definitely be improved.

Looks like it's just a yaml version of tm language. They probably got sick of XML and decided to write yaml first and compile to tmLanguage.
I believe sublime-syntax is more superior and does mostly everything tmLanguage does and more. If you intend to stick with Sublime for the foreseeable future like I do, we should just write sublime-syntax. The Elixir syntax is converted from the tmLanguage version, which is done by Sublime itself. The definition is not optimal, I already know and have made improvements to it. Let's make it better together!

azizk commented:
> The definition is not optimal, I already know and have made improvements to it. Let's make it better together!

Yeah, let's do so! :)

I added one commit to fix the issue of matching numbers in function names like `def 123() do end` for example. While I was at it, I added a variable which should make the regex clearer.

princemaple commented:
Thanks @azizk ! Merged.
